8261205: AssertionError: Cannot add metadata to an intersection type

vicente-romero-oracle opened the pull request:
Please review this PR which is adding a field to `JCVariableDecl`, the need I see for this is that when we declare a local variable using `var` as in:

```
import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target({ElementType.TYPE_USE, ElementType.LOCAL_VARIABLE})
@interface A {}

class Test {
    void t() {
        @A var c = g(1, 1L);
    }

    <X> X g(X a, X b) {
        return a;
    }
}
```
the only clue the compiler has to know that the local variable was declared using `var` is that field `JCVariableDecl::vartype` is null, but this field is set to an inferred type at some point during the compilation and from that point on there is no way to know for sure that the original AST had a `var` or not. Unfortunately there is code that keeps asking if the local variable was implicitly typed or not, after field `JCVariableDecl::vartype` has been written, and thus making uninformed decisions as in: TypeAnnotations.TypeAnnotationPositions::visitVarDef, where the compiler is implementing a portion of section: `9.7.4 Where Annotations May Appear` of the spec (JLS16). In particular the portion that defines how to deal with annotations applied to local variables declared with `var`.

The test case above is interesting because the same annotation works as a declaration annotation **and** as a type annotation. If it were only a type annotation then the compiler would have issued an error before getting to `TypeAnnotations.TypeAnnotationPositions::visitVarDef`, precisely at `Check::validateAnnotation`, see that `Check::getApplicableTargets` has a special case for local variables declared using `var`. But again given that the annotation in the example above is applicable as a type and as a declaration annotation, it will seep down to the metadata info associated to the local variable symbol, and that is fine. What is not kosher is that as the compiler has no clue that the AST was originally declared using `var` that it then tries to type-annotate the type of the local variable. This change is thus making sure that that information is preserved.

This patch is also renaming a method in `JCVariableDecl`, its previous name was `isImplicitlyTyped` which was basically checking if the `vartype` field was `null` and was renamed to `nullVarType`, `hasNullVarType` could be an option too, and a new method was added: `declaredUsingVar` which won't change its returned value during the lifetime of a JCVariableDecl AST.

TIA

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261205](https://bugs.openjdk.java.net/browse/JDK-8261205): AssertionError: Cannot add metadata to an intersection type


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to 0552cd4fd59dd2d75ef5312b3132b4b545f55e16


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4095/head:pull/4095` \
`$ git checkout pull/4095`

Update a local copy of the PR: \
`$ git checkout pull/4095` \
`$ git pull https://git.openjdk.java.net/jdk pull/4095/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4095`

View PR using the GUI difftool: \
`$ git pr show -t 4095`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4095.diff">https://git.openjdk.java.net/jdk/pull/4095.diff</a>

</details>

mcimadamore reviewed (commented):
What happens if we just use `declaredUsingVar` everywhere? Shouldn't that be the "right" way to do it?

vicente-romero-oracle commented:
> What happens if we just use declaredUsingVar everywhere? Shouldn't that be the "right" way to do it?

yep I also considered that but then realized that parameters of implicit lambdas also have its vartype set to null until the compiler infers a type for them, so the implicitly-typed set and the declared using `var` set have an intersection but are not the same, I preferred to keep the previous API for the different clients, just with a tweak in the method name, and add a new method to reflect the fact that there is a difference between being explicitly typed and being declared using `var`

vicente-romero-oracle commented:
I have uploaded a new commit, after an offline comment from Maurizio, he is correct that: given that lambda parameters also have a null type, they will be classified as "declared with var". This commit gets back to the original method name `JCVarDeclaration::isImplicitlyTyped` which will be `true` for local variables declared with `var` and for parameters of implicit lambdas.

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/MemberEnter.java:
```diff
@@ -44,7 +44,6 @@
 import static com.sun.tools.javac.code.Kinds.*;
 import static com.sun.tools.javac.code.Kinds.Kind.*;
 import static com.sun.tools.javac.code.TypeTag.TYPEVAR;
-import static com.sun.tools.javac.tree.JCTree.Tag.VARDEF;
```
this change have been introduced by intelliJ

mcimadamore reviewed (commented):
Looks better - one question: what about `var` with annotations and lambda parameters? E.g. you can probably construct a similar example involving a lambda parameter - is this fix enough, or would the code path be slightly different?

vicente-romero-oracle commented:
> Looks better - one question: what about `var` with annotations and lambda parameters? E.g. you can probably construct a similar example involving a lambda parameter - is this fix enough, or would the code path be slightly different?

good catch! I have modified the patch to include the case for lambda parameters declared with `var`, I also removed the initial test and added a new one that not only makes sure that the code compiles, it also checks that no type annotations are represented in the class file

mcimadamore commented on test/langtools/tools/javac/annotations/typeAnnotations/VariablesDeclaredWithVarTest.java:
```diff
@@ -0,0 +1,138 @@
+
+            class Test {
+                void kaa() {
+                    //@A var c = g(1, 1L);
```
Is this meant to be commented out?

mcimadamore replied:
Ah ok, I see this is fixed now

vicente-romero-oracle replied:
yep, thanks for the review

mcimadamore reviewed (commented):
Looks good - but I think there's an issue in the test?

mcimadamore reviewed (approved):
Thumbs up!

vicente-romero-oracle commented:
/integrate
